### PR TITLE
fix: render multiline local image markdown

### PR DESCRIPTION
**Pull request by joescars**
## Thinking Path

- Hermes image-generation providers emit the reported form `![alt]\n(/home/user/.hermes/cache/images/image.png)`.
- The previous revision correctly narrowed explicit schemes but no longer fixed that reported bare-cache-path form.
- The safe compatibility boundary is the shared Markdown image pass: recognize bare absolute paths only when they are image files beneath a `/cache/images/` path segment.
- Root-relative web URLs, protocol-relative URLs, traversal paths, and blank-paragraph continuations must remain inert/unmodified.

## What Changed

- Added a cache-scoped bare-path compatibility matcher in `static/ui.js`.
- Accepts one newline (or horizontal whitespace) between the alt text and target; blank paragraphs do not join.
- Normalizes an accepted bare cache path to `file://` and routes it through the existing session-authorized `api/media` flow.
- Leaves `/images/logo.png` and `//cdn.example/image.png` unchanged.
- Rejects traversal-shaped cache paths.
- Added the required rendering contract documentation to `docs/CONTRACTS.md`.
- Added regression tests for the positive case and all requested negative boundaries.
- No dependency, build-system, API, session, or authentication changes.

## Contract Routing

Task type: UI Markdown rendering bug fix.
Touched areas: `static/ui.js`, `tests/test_data_uri_images.py`, `docs/CONTRACTS.md`.
Relevant public docs:
- `AGENTS.md`
- `CONTRIBUTING.md`
- `docs/CONTRACTS.md`
- `docs/GUIDELINES.md`
- `TESTING.md`
Scope boundaries: the existing `/api/media` authorization remains unchanged; only Markdown recognition and documentation changed.

## Verification

The focused regression suite was run against the changed head:

```text
node --check static/ui.js

./scripts/test.sh tests/test_data_uri_images.py tests/test_renderer_js_behaviour.py
93 passed

git diff --check
passed
```

Coverage includes:

- Bare multiline Hermes cache path renders inline.
- Root-relative `/images/logo.png` remains unchanged.
- Protocol-relative `//cdn.example/image.png` remains unchanged.
- Blank-line continuation does not join paragraphs.
- Cache traversal remains inert.
- Existing explicit `file://`, HTTPS, data-image, and local media behavior remains covered.

## UI Evidence

Before: the reported screenshot shows the alt text and local path rather than the generated image. The regression fixture uses that exact output shape. A live after screenshot is not available in this environment because the reporter's authenticated WebUI session and generated image are not running here. Maintainer manual verification should confirm desktop and narrow/mobile rendering after the PR updates.

## Risks / Follow-ups

- The cache matcher is intentionally path-shape based and relies on the existing `/api/media` server-side allowlist for final authorization.
- The PR does not expose arbitrary filesystem paths or reinterpret root-relative web assets.
- The issue remains open until maintainer review and merge.

## Release Note

Render Hermes-generated local cache images when Markdown places the bare cache path on the next line, without broadening arbitrary local-path handling.

## Model Used

OpenAI GPT-5.6 Luna via Hermes Agent, using repository inspection, patch tools, Node syntax validation, the repository test runner, Git, and GitHub CLI. No credentials or private data were added.

### CODE REVIEW

**Comment by joescars** [comment]
Closes #7634

**Comment by joescars** [comment]
Contributor-guide verification update: the required full runner completed with 15,084 passed, 482 unrelated pre-existing failures, 186 skipped, 2 xfailed, 2 xpassed, 13 warnings, and 45 subtests passed. The failures are outside the changed files and cover title sanitization, turn-journal callsites, UIUX docs theme contracts, gateway backend, notes sources, and worktree removal. The focused image test (12 passed) and neighboring renderer suite (75 passed) remain green. I have documented this limitation in the PR body rather than claiming the full suite passed.

**Review by nesquena-hermes — CHANGES_REQUESTED** [review]
Thanks — the core bug is real and worth fixing. Hermes image-generation providers do emit `![alt]\n(/path/to/cached.png)` with the path on the next line, and on master that renders as literal alt text instead of an image. Your fix does resolve it (verified against the real `renderMd()`).

But the same one-line regex change does three things at once, and two of them are wider than the fix needs. I'd like those separated before this ships.

## The blocker: bare absolute paths become clickable download links

The regex adds `|\/` to the accepted URL schemes, so **any** `x` in model output now resolves to `/api/media`. I probed this against the real renderer:

| input | master | this PR |
|---|---|---|
| `x` | inert literal text | `<a class="msg-media-link" href="api/media?path=%2Fetc%2Fshadow&download=1">📎 shadow</a>` |
| `x` | inert literal text | `📎 id_rsa` download link |
| `x` | inert literal text | `📎 .env` download link |
| `x` | inert literal text | `📎 passwd` download link |

**This is not a disclosure vulnerability** — I confirmed the server holds the line: `/api/media?path=/etc/shadow` returns **403 `Path not in allowed location`**, and the session-token allow-list (`_session_media_token_allows_path`) means assistant-emitted tokens can't mint entries. Defense-in-depth works.

It is still a real regression in posture. Model output is untrusted input, and this turns an inert string into a rendered, clickable, `download=`-attributed link pointing at sensitive host paths. A user who clicks gets an error, but the affordance itself is misleading and is exactly the kind of thing that reads as a phishing surface in a screenshot. Master deliberately required an explicit scheme (`https://`, `file://`, `data:image/`) for that reason.

Also note `x` — a protocol-relative URL — now matches the `/` branch and is rewritten to `file:///evil.test/a.png` → `api/media?path=%2F%2Fevil.test%2Fa.png`, which is not what either reading of that input intends.

## Requested change

Keep the multiline fix; drop the scheme widening. The newline tolerance is the actual bug:

```js
/!\[([^\]]*)\]\s*\(\s*((?:https?:\/\/|file:\/\/|data:image\/)[^\)]+?)\s*\)/g
```

That alone makes `![cat]\n(file:///path/x.png)` work. If the generators genuinely emit a bare path with no `file://` prefix, that's worth fixing too — but as a deliberate, separately-reviewed change with a path allow-list (e.g. confined to the images cache dir) rather than accepting any `/…`, and with the protocol-relative case explicitly excluded.

## Smaller notes

- The `\s*` additions also make `x` match where master didn't. Harmless in isolation, but it's a third behavior change riding along in the same line.
- The new test only covers the happy path. Whichever shape you land, please add negative cases — `x`, `x`, and a traversal path — asserting they do **not** become links.

Tests on the rebased head are green (1248 passed) — the gap is coverage, not failures.

Happy to re-gate as soon as you push. If you'd rather I split it myself into the narrow multiline fix now and file the bare-path question separately, say the word and I'll do that with you credited.

**Comment by joescars** [comment]
Addressed the requested changes in commit `5c776e29` (rebased onto current `master`):

- Kept newline/whitespace tolerance.
- Removed the broad bare-`/` matcher and automatic `file://` conversion.
- Added negative tests for `/etc/shadow`, `//evil.test/a.png`, and traversal.
- Updated the PR description to document the narrowed behavior.

Verification:
- `tests/test_data_uri_images.py`: 15 passed
- image tests plus `tests/test_renderer_js_behaviour.py`: 92 passed

**Comment by joescars** [comment]
Addressed the requested changes in commit `5c776e29` (rebased onto current `master`):

- Kept newline/whitespace tolerance.
- Removed the broad bare-`/` matcher and automatic `file://` conversion.
- Added negative tests for `/etc/shadow`, `//evil.test/a.png`, and traversal.
- Updated the PR description to document the narrowed behavior.

Verification:
- `tests/test_data_uri_images.py`: 15 passed
- image tests plus `tests/test_renderer_js_behaviour.py`: 92 passed

**Comment by joescars** [comment]
Addressed the latest review feedback in commit `2ec8143a`:

- Bare absolute paths are recognized only when they are image files under a `/cache/images/` segment.
- Root-relative web images such as `/images/logo.png` remain unchanged.
- Protocol-relative images such as `//cdn.example/image.png` remain unchanged.
- Traversal-shaped paths remain inert.
- Only one line break is accepted between the Markdown image label and target; blank paragraphs do not join.
- Added the rendering contract documentation to `docs/CONTRACTS.md`.
- Added regression tests for the positive and negative cases.

Verification on the pushed head:

- `node --check static/ui.js` passed.
- `./scripts/test.sh tests/test_data_uri_images.py tests/test_renderer_js_behaviour.py`: 93 passed.
- `git diff --check` passed.

**Comment by nesquena-hermes** [comment]
## Summary

I re-read the full three-file diff at exact head `2ec8143a8d677ee6fc519f1e8ebc4b58270b4e71`, the complete renderer helper path, the updated tests, and the server media authorization path. The broad bare-absolute-path problem from my prior review is mostly fixed: ordinary root-relative and protocol-relative URLs stay inert, and the new matcher is limited to `cache/images` plus raster extensions. One normalization-order gap remains, however: traversal is checked before the accepted path is URL-decoded, while the downstream media helper decodes it before requesting the file.

## Code reference

At `static/ui.js:8050-8056`, the new path gate operates on the raw Markdown target:

```js
const hasTraversal=/(^|\/)\.\.(?:\/|$)/.test(url);
const bare=_bareHermesImageCacheRe.test(url)&&!hasTraversal;
if(hasTraversal||(!/^(?:https?:\/\/|file:\/\/|data:image\/)/i.test(url)&&!bare))return `![${alt}](${url})`;
const normalized=bare?`file://${url}`:url;
```

But `_inlineMediaHtmlForRef()` later runs `decodeURIComponent(u.pathname)` at `static/ui.js:2763`. Therefore a target such as `/x/cache/images/%2e%2e%2fother.png` contains no raw `/../`, satisfies the filename/extension matcher, and is only then decoded into `/x/cache/images/../other.png`. A backslash-shaped segment has the same normalization concern on URL implementations that treat backslashes as separators.

There is a second, smaller mismatch in the same regex: it accepts `~/...`, then constructs `file://~/...`. In URL syntax `~` becomes the host and `u.pathname` drops it, so the resulting media request points at a different absolute path. The PR body and docs promise bare absolute paths, not tilde paths, so this branch is both untested and outside the stated contract.

The server still resolves and re-authorizes the final target at `api/routes.py:20722-20805`, so this is not evidence of an arbitrary-file disclosure. It does break the client-side promise that traversal-shaped targets remain inert and can turn an encoded sibling path under an allowed root into a real media request.

## Diagnosis / recommendation

Please canonicalize before applying the compatibility allowlist. Decode exactly once with failure returning inert text, normalize or reject backslashes, then run the `cache/images` shape and `..` segment checks on that decoded value. Alternatively, reject percent and backslash characters in this narrow bare-path form. Remove the `~/` branch unless it is given explicit, tested expansion semantics.

## Test plan

Extend `tests/test_data_uri_images.py:133-165` with percent-encoded traversal, backslash traversal, malformed percent encoding, and `~/cache/images/...`; each should remain inert and produce no `api/media` request. Keep the existing positive multiline absolute-path case as the control.

This was read-only inspection; no contributor-authored code was executed.

**Comment by joescars** [comment]
Addressed the maintainer’s latest review in commit `1e1c4926`.

## Changes made

- Canonicalize bare cache-path candidates with one `decodeURIComponent()` before applying the compatibility allowlist.
- Reject malformed percent escapes by treating the reference as inert text.
- Reject traversal after decoding, including encoded `..` segments such as `%2e%2e%2f`.
- Reject backslash-shaped paths and encoded backslashes before they reach URL/path normalization.
- Removed the unsupported `~/` compatibility branch; only absolute `/...` paths can qualify.
- Preserve the existing cache/image extension and `/cache/images/` restrictions.
- Added regression tests for:
  - Encoded traversal
  - Backslash traversal
  - Malformed percent escapes
  - Tilde paths
  - The existing valid multiline bare cache path remains covered

## Verification

```text
node --check static/ui.js
passed

./scripts/test.sh tests/test_data_uri_images.py tests/test_renderer_js_behaviour.py
96 passed

git diff --check
passed
```

The change was pushed to the PR branch. The PR now contains the latest maintainer-requested normalization and negative-coverage fixes while preserving the original MAI cache-image rendering behavior.

**Comment by nesquena-hermes** [comment]
## Summary

I re-read the full three-file diff at exact head `1e1c4926`, including the revised renderer pass, every added regression, and the unchanged local-media helper and server authorization path. The single-encoded traversal from my prior review is now rejected, and tilde/backslash handling is narrowed correctly. The path is still decoded twice across the real pipeline, however, so a double-encoded traversal remains accepted by the new allowlist and becomes traversal only inside `_inlineMediaHtmlForRef()`.

## Code reference

The new gate at `static/ui.js:8052-8068` decodes once, validates that intermediate string, and then wraps it in another `file://` URL:

```js
const decodedBare=_decodeBareHermesImagePath(url);
const bare=decodedBare!==null&&_bareHermesImageCacheRe.test(decodedBare);
const explicit=/^(?:https?:\/\/|file:\/\/|data:image\/)/i.test(url);
if(!explicit&&!bare)return `![${alt}](${url})`;
const normalized=bare?`file://${decodedBare}`:url;
```

The downstream helper then parses that URL and calls `decodeURIComponent(u.pathname)` at `static/ui.js:2762-2765`. For `/x/cache/images/%252e%252e%252fother.png`, the first decode produces `/x/cache/images/%2e%2e%2fother.png`; that still matches `_bareHermesImageCacheRe`, contains no literal dot-dot segment, and is wrapped as a file URL. The second decode produces `/x/cache/images/../other.png` and routes it to `api/media`.

The server still resolves and authorizes the final path at `api/routes.py:20784-20808`, so this is not evidence of arbitrary-file disclosure. It does violate the new client contract in `docs/CONTRACTS.md:24-34`: encoded traversal is supposed to remain inert rather than become a media request.

## Diagnosis / recommendation

For this deliberately narrow compatibility form, the safest contained rule is: decode once, then reject the candidate if the decoded value still contains `%` before running the cache-path regex. That permits ordinary encoded spaces after one decode while preventing any second decoding layer from changing path structure. A bounded repeated-decode loop is weaker because another encoding layer can always sit beyond the bound.

## Test plan

Extend the negative coverage beside `tests/test_data_uri_images.py:160-177` with double-encoded slash traversal (`%252e%252e%252f`) and double-encoded backslash traversal (`%252e%252e%255c`). Assert both remain literal Markdown, produce no image, and produce no `api/media` URL. Keep the current valid multiline path as the positive control.

This was read-only inspection; no contributor-authored code was executed.

**Comment by joescars** [comment]
Addressed the maintainer’s latest double-encoding review in commit `b9fc58fb`.

## Changes

- Bare cache-path candidates are still decoded once before validation.
- Any residual `%` after that decode is now rejected as inert text. This prevents a second downstream decode from changing the path structure.
- Added regression tests for:
  - Double-encoded slash traversal: `%252e%252e%252f`
  - Double-encoded backslash traversal: `%252e%252e%255c`
- Preserved the valid multiline absolute cache-path control case.
- Preserved the existing single-encoded traversal, backslash, malformed-escape, tilde, root-relative, protocol-relative, and blank-paragraph tests.

## Verification

```text
node --check static/ui.js
passed

./scripts/test.sh tests/test_data_uri_images.py tests/test_renderer_js_behaviour.py
97 passed

git diff --check
passed
```

The fix is pushed to the PR branch at the latest head `b9fc58fb`. The residual-percent rejection is deliberately limited to the bare Hermes cache-path compatibility form; explicit `https://`, `file://`, and `data:image/` handling remains unchanged.

**Comment by joescars** [comment]
Addressed the latest automated-review comment in commit `003a6213`.

## Finding addressed

The previous residual-percent guard rejected legitimate generated cache filenames such as:

```text
/home/user/.hermes/cache/images/photo%25done.png
```

That is now supported again.

## Safety retained

- Percent-encoded path separators (`%2f`, `%5c`) are rejected.
- Encoded traversal components remain rejected.
- Double-encoded traversal remains rejected because residual encoded traversal delimiters are rejected before the downstream decode.
- Backslash-shaped paths, malformed escapes, tilde paths, root-relative URLs, protocol-relative URLs, and blank-paragraph joins remain inert.
- Accepted paths are re-encoded into the `file://` URI before entering the existing media helper, so a literal percent in a valid filename is preserved correctly.

## Test coverage

Added a positive regression test for a percent-containing cache filename and retained the double-encoded traversal tests.

Verification:

```text
node --check static/ui.js
passed

./scripts/test.sh tests/test_data_uri_images.py tests/test_renderer_js_behaviour.py
98 passed

git diff --check
passed
```

The fix is pushed to the PR branch at head `003a6213`.

**Comment by joescars** [comment]
@nesquena-hermes  - are you able to provide a final review?

**Review by nesquena-hermes — CHANGES_REQUESTED** [review]
## Re-gate at `c89c68142307`: the path narrowing fixes the earlier blocker; two renderer edge cases remain

Thanks @joescars. The narrowing works: bare paths outside `cache/images`, `//host`, traversal, encoded separators, tilde and backslash forms all stay inert, and the related tests pass on current master (rebased clean). A Codex re-gate of the real `renderMd()` found two remaining problems. I reproduced both.

### 1. [SILENT, reproduced] Inline code no longer protects the path

The outer image pass runs after inline code is restored, so it rewrites content inside `<code>`. Using your own `test_data_uri_images.py` driver:

| input | master | this PR |
|---|---|---|
| `` `x` `` | literal code | `<code>` |
| `` - `x` `` (list) | literal code | image inside `<li><code>` |

Code spans are how users and models show markdown *as text*, and this also triggers an `api/media` request.

### 2. [BRICK-class, reproduced] Malformed Unicode throws out of `renderMd()`

A lone UTF-16 surrogate in the cache path, which JSON-loaded message text can contain via a `\ud800` escape, makes `encodeURIComponent` throw `URIError: URI malformed` from `_bareHermesImageFileUri`. I reproduced the throw by running the real `renderMd()` in a Node VM. Master renders the same input as text. The caller effect is inferred: `renderMessages()` clears the transcript before rendering each message, so a throw here can leave a restored session blank.

### Fix spec

1. Wrap the encoding in `try/catch` and return the original markdown inert on failure. Add a renderer test with a lone surrogate. It has to be built inside the Node driver (for example `'\uD800'` in JS), because piping it through stdin re-encodes it to U+FFFD.
2. Keep code spans stashed through the outer image pass (or skip matches inside `<code>`). Test plain, list and table code spans. Existing images, `https`, `data:image`, `file://` and table-cell images must keep rendering.

This is a visible rendering change, so after these land it goes through a screenshot and UX pass before ship. I'll re-gate the new head first.

**Comment by joescars** [comment]
Implemented the latest maintainer re-gate findings on the PR branch in commit `e46b109517006853f26f68040ad0b9efa102b029`.

- Preserved inline-code image text through the outer image rewrite in paragraphs, list items, and table cells.
- Made cache-file URI encoding fail closed on malformed Unicode, leaving the Markdown inert rather than throwing from `renderMd()`.
- Added real-renderer regression cases for those four inputs, plus explicit HTTPS, `file://`, and `data:image/` table-image controls.

Verification on this commit: `node --check static/ui.js`; `./scripts/test.sh tests/test_data_uri_images.py tests/test_renderer_js_behaviour.py tests/test_blockquote_rendering.py -q` (126 passed); `git diff HEAD --check`. The new inline-code and lone-surrogate cases were also run against the pre-fix renderer and all four failed as expected.

Contract routing: chat Markdown rendering; `docs/CONTRACTS.md`; safe behavior remains bounded to the cache-image compatibility path. Generated Chrome screenshots were captured at 1440×900 and 375×812 before/after; this renderer-path-only fix changes no page layout. The full suite started against the current local snapshot but was interrupted after early failures; results are not claimed. Maintainer review can re-gate the pushed head.

File: static/ui.js [hunk 1]
```diff
@@ -8087,7 +8087,44 @@ function renderMd(raw){
...
+    if(codeRange&&offset>=codeRange[0]&&offset<codeRange[1])return match;
+    const url=String(rawUrl||'').trim();
+    const decodedBare=_decodeBareHermesImagePath(url);
+    const bare=decodedBare!==null&&_bareHermesImageCacheRe.test(decodedBare);
+    const explicit=/^(?:https?:\/\/|file:\/\/|data:image\/)/i.test(url);
+    if(!explicit&&!bare)return `![${alt}](${url})`;
+    const normalized=bare?_bareHermesImageFileUri(decodedBare):url;
+    if(normalized===null)return `![${alt}](${url})`;
+    return(typeof _mdImageHtml==='function')?_mdImageHtml(alt,normalized):`<img src="${normalized.replace(/"/g,'%22')}" alt="${esc(alt)}" class="msg-media-img" loading="lazy">`;
+  });
```
[the remark this answers is not in the document]
Comment (joescars) [reply static/ui.js]: Addressed in commit `884181a9c1235a41e9119f0b0b98322cc2f546ae`.

Added before/after screenshots from the real Hermes WebUI at desktop (1440×900), narrow (900×900), and mobile (390×844), and linked them in the PR description and `docs/PR-7635-UI-EVIDENCE.md`.

Verification used the same two-message transcript and local sample image in isolated, localhost-only previews. Chrome 154 DOM checks at each viewport found four image nodes before (the intended image plus three code-span rewrites) and one after (the intended image only); the three code examples remained literal at the fixed head. All six screenshot URLs and the evidence note are reachable from the pushed head. No app styling or layout changed. The new Greptile check is currently re-running.

**Review by nesquena-hermes — CHANGES_REQUESTED** [review]
## Re-gate at `2e0e5dc8b425`: both earlier findings are fixed; raw `<code>` in a paragraph still renders a live image

Thanks @joescars. **Both findings from my last review are resolved.** Backtick code spans keep the path inert in paragraphs, list items and table cells, and malformed Unicode no longer throws out of `renderMd()`. The 126 focused renderer tests pass.

### Silent: raw `<code>` HTML still turns the path into a live image (`static/ui.js:8110`)
I ran the real `renderMd()` through your test driver on the same input for both builds:

| Input | master | this head |
|---|---|---|
| ``See `x` for the syntax.`` | text | text ✅ |
| `See <code>x</code> for the syntax.` | text | **``** |

Raw `<code>` is rewritten into backticks around `static/ui.js:7830`, *after* the initial code stash, so the new guard never sees it as code. Models often use raw `<code>` when they explain markdown syntax, so an "here's how to embed an image" answer would load a local file from the media endpoint.

**Fix:** protect raw `<code>…</code>` (and the backtick form it's converted to) through the outer image pass: stash it before the rewrite, or run the rewrite before the raw-`<code>` conversion. Please add real-renderer cases for raw `<code>` and entity-escaped `&lt;code&gt;` in a plain paragraph, asserting no `<img` and no `api/media`.

Minor: the 6 PNGs under `evidence/pr-7635/` would be added to the repo permanently. Please attach them to the PR description instead and drop them from the branch.

I'll re-gate the next head.

**Comment by joescars** [comment]
Addressed the latest maintainer re-gate finding in commit `a23e5eb9`.

- The outer image rewrite now also protects backtick spans generated from raw and entity-escaped `<code>…</code>` markup earlier in `renderMd()`.
- Added real-renderer regressions for raw `<code>` and entity-escaped `&lt;code&gt;`, asserting no `` or `api/media` output. Existing inline, list, and table code-span coverage remains.

Verification:
- `./scripts/test.sh tests/test_data_uri_images.py tests/test_renderer_js_behaviour.py tests/test_blockquote_rendering.py -q`: 128 passed
- `node --check static/ui.js`: passed
- `git diff --check`: passed

Contract routing remains chat Markdown rendering (`docs/CONTRACTS.md`); no layout change. Responsive before/after evidence from the prior head remains in the PR description/evidence note.

**Comment by joescars** [comment]
Addressed the maintainer re-gate finding on raw and entity-escaped HTML code in commit `a23e5eb935a7e17b244b6fca8511d4501de0eeb4`.

- Protect backtick spans introduced when raw `<code>…</code>` is converted earlier in `renderMd()`, and include those ranges in the existing outer image-pass exclusion.
- Added real-renderer regression coverage for raw `<code>` and entity-escaped `&lt;code&gt;` image syntax. Both assert the source remains literal and emits neither `` nor an `api/media` URL.
- The regression cases failed on the prior renderer; they pass with this change.

Verification on the pushed head:
- `node --check static/ui.js` passed.
- `HERMES_WEBUI_TEST_STATE_DIR=/home/joe/Code/hermes-webui/.git/test-state-7635 ./scripts/test.sh tests/test_data_uri_images.py tests/test_renderer_js_behaviour.py tests/test_blockquote_rendering.py -q` — 128 passed.
- `git diff --check` passed.

The existing responsive before/after captures remain unchanged; this is a renderer-only fix with no layout or styling changes. The PR body still documents the pre-existing incomplete full-suite run.

**Review by nesquena-hermes — CHANGES_REQUESTED** [review]
## Re-gate at `e960a13153d3`: raw `<code>` images are inert now, but a real image *between* two raw `<code>` spans stops rendering

Thanks @joescars. **The raw-`<code>` finding is fixed.** `See <code>x</code>` now renders literal text with no `` or `api/media` (checked live in the browser `renderMd()` on this head). Backtick code, plain images, and `` `a` and x and `b` `` are all unchanged from master.

### Silent: the new backtick-pairing pass swallows an image between two code elements (`static/ui.js:8116-8125`)
The second pass collects `` `[^`\n]+` `` ranges over the string *after* raw `<code>` was turned into backticks. When a code element itself contains a backtick, the pairing jumps from one element to the next, and the range covers the image in between. The image pass then skips it.

Same input, in the browser `renderMd()` on a live seeded server (`<code>first`last</code> x <code>end</code>`):
- **master `9a00b652c`:** `` `first`last`  `end` ``
- **this head:** `` `first`last` !<a href="https://example.test/a.png" …>x</a> `end` `` (the image becomes a link; Codex also saw `data:image/` become literal Markdown and `file://` become a link)

(Note: the Node `_render` driver in `tests/test_data_uri_images.py` renders a link for this input on master *and* the PR, so it can't catch this. Please check it through the browser path, or a driver that matches it.)

**Fix:** drop the global post-conversion backtick scan. Record each raw `<code>…</code>` region by its own start and end at the point it's converted (`static/ui.js:~7830`), and carry those exact ranges into the image pass. That way only text actually inside a raw `<code>` element is suppressed. Please add a regression for the input above (expect ``) alongside the existing inert-cache case.

**Comment by joescars** [comment]
Addressed the interleaved raw-code finding in commit `2436bf6e2f74954eb5e72db10f8f758471f4f574`. Raw `<code>` elements now stay opaque placeholders through Markdown image/link processing and are restored as escaped code afterward, so backticks inside one element cannot pair across another element and suppress an intervening image. Added a regression using the reported `<code>first`last</code> x <code>end</code> shape. The regression failed before the fix and passes now. Verification: `./scripts/test.sh tests/test_data_uri_images.py tests/test_renderer_js_behaviour.py tests/test_blockquote_rendering.py -q` (129 passed), `node --check static/ui.js`, and `git diff --check` all passed.

**Review by nesquena-hermes — CHANGES_REQUESTED** [review]
## Re-gate at `2436bf6e2f74`: images between raw `<code>` render again; a backtick inside raw `<code>` now leaks a stash token

Thanks @joescars. The finding from my last review is **fixed**: an image between two raw `<code>` elements renders again, and image syntax inside raw or entity-escaped code stays inert. Codex confirmed it in a Chromium harness running the production `renderMd()`, 129 renderer tests pass, and the PR's 31 tests pass rebased onto current master. Plain raw code also improved: `Run <code>npm test</code> and <code>npm run build</code>` now renders two real `<code>` elements, where master shows literal backticks.

I ran `renderMd()` in the real browser (seeded server, this PR vs master) on a few more inputs. One shape regresses:

### CORE: a backtick inside a raw `<code>` element leaks a stash token and drops text (`static/ui.js`, the new `rawCodeStash` restore)
Input (normal wording when explaining Markdown):

```
Use <code>`</code> for inline code and <code>```</code> for fences.
```

| | rendered text | `<code>` contents |
|---|---|---|
| master | ``Use `</code> for inline code and <code>``` for fences.`` | 1 element (mangled, but no text lost) |
| this PR | ``Use F0`` for fences.`` | `F0```: a leaked `\x00F0\x00` token; "for inline code and" is gone |

A second shape loses an image the PR is meant to render: `` Type <code>`</code> then see i and `x`. `` renders the image on master and drops it on this PR.

What seems to happen: the restored raw-code contents still carry literal backticks, a later inline-code pass pairs them across elements, and its own stash token is created after that stash's restore has run. Master already mangles these inputs, but it keeps all the text and never leaks a token. This PR loses content, which is worse.

**Fix:** make restored raw-code contents inert to the later backtick passes, e.g. encode `` ` `` as `&#96;` when restoring `rawCodeStash`, or restore the raw-code stash after every backtick/inline-code pass. Please add browser-level regressions for the two inputs above. The Node `_render` driver in `tests/test_data_uri_images.py` didn't catch this, which is the same gap as last round.

Browser probe: `renderMd()` called in-page on a seeded server with `probe7635-bt.js` (four inputs; the two above differ, the other two match or improve on master).

**Comment by joescars** [comment]
Addressed the latest maintainer browser re-gate findings in commit `ae2619632a81f08ddee9991d37947d076d7e3424`.

### Fix

Raw `<code>` elements are now stashed immediately after entity decoding, before inline-backtick processing. Their contents are restored escaped after Markdown processing, with literal backticks encoded so later renderer stages cannot reinterpret them.

### Regression coverage

Added both requested cases to `tests/browser_smoke.py`, running the production `renderMd()` in Chromium:

- Separate raw `<code>` elements containing single and triple backticks preserve all text and produce the expected code elements without leaking a stash token.
- A backtick-containing raw code element before a Markdown image and inline code preserves the rendered image and adjacent content.

### Verification

- Focused renderer suites: **129 passed**
- `node --check static/ui.js`: passed
- `python3 -m py_compile tests/browser_smoke.py`: passed
- `git diff --check`: passed
- `HERMES_WEBUI_PORT=8796 .venv/bin/python tests/browser_smoke.py`: passed, including the Chromium renderer regressions

The PR remains open and behind `master`; CI is re-running on this head.

**Comment by joescars** [comment]
## Greptile request addressed: fenced code remains literal

Commit [`14e102b4`](https://github.com/nesquena/hermes-webui/commit/14e102b4) moves raw `<pre>` and inline `<code>` shielding until after fenced Markdown has been extracted into opaque placeholders. This prevents `<code>…</code>` examples inside a fenced block from being stashed and later restored as active nested markup, while preserving the existing raw-code/backtick protection outside fences.

Added a Chromium regression that passes ```` ```html
<code>foo</code>
``` ```` through production `renderMd()` and verifies the literal source text is shown in a code element.

**Verification**

- `./scripts/test.sh tests/test_data_uri_images.py tests/test_renderer_js_behaviour.py tests/test_blockquote_rendering.py -q` — 129 passed
- `node --check static/ui.js` — passed
- `python3 -m py_compile tests/browser_smoke.py` — passed
- `HERMES_WEBUI_PORT=8796 .venv/bin/python tests/browser_smoke.py` — passed in Chromium; all pages loaded with zero console errors
- `git diff --check` — passed

**Review by nesquena-hermes — CHANGES_REQUESTED** [review]
## Re-gate at `14e102b426a8`: the backtick leak is fixed; inline code that *shows* an HTML tag now loses the tag

Thanks @joescars. The last round's leak is fixed. I re-ran my live probe in a real browser with the production `renderMd()`: all four raw-`<code>`-containing-a-backtick cases now render correctly with no NUL, and `<code>` inside a ```fence``` stays literal. The claim itself works too: `![alt]\n(/…/.hermes/cache/images/x.png)` becomes an `api/media` image where master shows the literal text. These stay inert: `/etc/passwd`, `../` traversal, root-relative, protocol-relative, blank-paragraph, inline-code and a `.sh` extension.

**One regression is left (found by Codex, reproduced live).** Moving the raw-tag stash ahead of inline-backtick parsing means that an inline code span that *contains* a raw tag now renders the tag as live HTML. That's the usual way to show HTML syntax in prose.

| Input | This PR | Master |
|---|---|---|
| ``Explain `<code>npm test</code>` syntax.`` | "Explain **npm test** syntax." (nested `<code><code>`, tags gone) | "Explain `<code>npm test</code>` syntax." |
| ``Use `<pre>block</pre>` for preformatted text.`` | an empty `<code>` + a stray `block` element | "Use `<pre>block</pre>` …" |
| ``- Wrap with `<code>x</code>` `` (list item) | "Wrap with x" (nested) | literal tags |
| table cell ``| `<code>x</code>` |`` | "x" (nested) | literal tags |

Assistants write this constantly when explaining HTML, JSX or markdown itself, so this is a silent content loss on ordinary messages.

**Fix (Codex's suggestion, which matches what I'd do):** run the raw-tag stash and inline-backtick spans **in source order**. A raw `<code>`/`<pre>` *inside* a backtick span must stay escaped text belonging to that span; only a standalone raw `<code>`/`<pre>` should be stashed as opaque. One way: while scanning, skip over backtick spans before looking for raw tags (as the fence pass already makes fences opaque). Please add production-renderer regressions (`tests/browser_smoke.py`, as you did for the last round) for the four rows above.

I'm also noting this in case it matters: a user-typed `\x00RC0\x00` renders the stashed content of the message's first raw `<code>`. That's pre-existing stash-design behavior (master does the same with `\x00R0\x00` for raw `<pre>`, and `\x00F0\x00` renders "undefined" on master). It only reads the same message's already-escaped text, so it's not blocking.

**Comment by joescars** [comment]
## Addressed: raw HTML examples inside inline code

Commit [`e4d3c0e6`](https://github.com/nesquena/hermes-webui/commit/e4d3c0e6) updates the renderer to scan inline backtick spans and raw `<pre>`/`<code>` elements in source order after fenced blocks are protected. When a backtick span starts first, HTML tags inside it remain escaped source text; standalone raw code elements are still shielded independently.

Added production-renderer Chromium regressions for the four reported cases: `<code>` and `<pre>` examples in prose, `<code>` in a list item, and `<code>` in a table cell. The tests assert the literal tag text and the expected block/cell parent.

**Verification**

- Focused renderer suites: 129 passed
- `node --check static/ui.js`: passed
- `python3 -m py_compile tests/browser_smoke.py`: passed
- Chromium browser smoke with production `renderMd()`: passed; all pages loaded with zero console errors
- `git diff --check`: passed
